File: libs/core/src/symbols/symbol.cpp
/**
 * @file symbol.cpp
 * @author Francesco Corso
 * @date 8/2/2026
 * @brief
 */
module;

#include <libassert/assert-macros.hpp>

module prism.core:symbols.symbol.impl;

import :symbols.symbol;
import :symbols.assembly_symbol;
import :symbols.namespace_symbol;
import :symbols.named_type_symbol;
import :compilation;
import :diagnostics.diagnostic_bag;

namespace prism
{

    DeclaredVisibility Symbol::declared_visibility() const
    {
        // TODO: For now symbol visibility does not apply but eventually it will
        return DeclaredVisibility::not_applicable;
    }

    bool Symbol::is_implicitly_declared() const noexcept
    {
        return false;
    }

    Optional<Location> Symbol::try_get_first_location() const
    {
        const auto locations = this->locations();
        if (locations.empty())
            return std::nullopt;

        return locations.front();
    }

    Location Symbol::first_location() const
    {
        auto location = try_get_first_location();
        if (!location.has_value())
            throw InvalidStateException{"Symbol has no location"};

        return *std::move(location);
    }

    Location Symbol::first_location_or_none() const
    {
        return try_get_first_location().value_or(no_location);
    }

    Optional<const AssemblySymbol &> Symbol::containing_assembly() const
    {
        return containing_symbol().and_then([](const Symbol &symbol) { return symbol.containing_assembly(); });
    }

    Optional<const NamespaceSymbol &> Symbol::containing_namespace() const
    {
        for (auto container = containing_symbol(); container.has_value(); container = container->containing_symbol())
        {
            auto ns = container->as<NamespaceSymbol>();
            if (ns.has_value())
                return ns;
        }

        return std::nullopt;
    }

    Optional<const TypeSymbol &> Symbol::containing_type() const
    {
        auto container = containing_symbol();
        auto containing_type =
            containing_symbol().and_then([](const Symbol &symbol) { return symbol.as<NamedTypeSymbol>(); });
        if (container.value_ptr() == containing_type.value_ptr())
        {
            return containing_type;
        }

        return container->containing_type();
    }

    bool Symbol::is_defined_in_source_tree(const SyntaxTree &tree, Optional<TextSpan> defined_within) const
    {
        auto declaring_references = declaring_syntax_references();
        if (is_implicitly_declared() && declaring_syntax_references().empty())
        {
            return containing_symbol().value().is_defined_in_source_tree(tree, defined_within);
        }

        for (auto &ref : declaring_references)
        {
            if (is_defined_in_source_tree(ref, tree, defined_within))
                return true;
        }
        return false;
    }

    bool Symbol::is_defined_in_source_tree(const SyntaxReference &reference,
                                           const SyntaxTree &tree,
                                           Optional<TextSpan> defined_within)
    {
        return &reference.tree() == &tree &&
               (!defined_within.has_value() || reference.span().intersects_with(defined_within.value()));
    }

    Optional<const Compilation &> Symbol::declaring_compilation() const
    {
        return containing_assembly().and_then([](const AssemblySymbol &assembly)
                                              { return assembly.declaring_compilation(); });
    }

    void Symbol::add_declaration_diagnostics(const DiagnosticBag &diagnostics) const
    {
        const auto compilation = declaring_compilation();
        DEBUG_ASSERT(compilation.has_value());
        CompilationInternal::get_declaration_diagnostics(*compilation).add_range(diagnostics);
    }

    bool Symbol::needs_completion() const noexcept
    {
        return false;
    }

    void Symbol::force_complete(const Optional<SourceLocation> &location, const Optional<SymbolPredicate> &filter) const
    {
        DEBUG_ASSERT(!needs_completion(), "Need to override force_complete");
    }

    bool Symbol::is_complete(CompletionPart part) const noexcept
    {
        DEBUG_ASSERT(!needs_completion(), "Need to override is_complete");
        return true;
    }

    void Symbol::force_complete_member_conditionally(const Optional<SourceLocation> &location,
                                                     const Optional<SymbolPredicate> &filter,
                                                     const Symbol &member)
    {
        if ((!location.has_value() || member.is_defined_in_source_tree(location->tree(), location->source_span())) &&
            (!filter.has_value() || (*filter)(member)))
        {
            member.force_complete(location, filter);
        }
    }
} // namespace prism

File: src/Prism.Core/Symbols/Symbol.cs
using System.Collections.Immutable;
using System.Diagnostics;
using Prism.Core.Compiling;
using Prism.Core.Declarations;
using Prism.Core.Diagnostics;
using Prism.Core.Syntax;
using Prism.Core.Text;

namespace Prism.Core.Symbols;

public abstract class Symbol
{
    private protected Symbol(string name, Symbol? containingSymbol = null)
    {
        Name = name;
        ContainingSymbol = containingSymbol;
    }

    public string Name { get; }

    // TODO: For now symbol visibility does not apply but eventually it will
    public virtual DeclaredVisibility DeclaredVisibility => DeclaredVisibility.NotApplicable;

    public abstract ImmutableArray<Location> Locations { get; }

    public Location FirstLocation()
    {
        return TryGetFirstLocation()
            ?? throw new InvalidOperationException("Symbol has no locations");
    }

    public Location? TryGetFirstLocation()
    {
        return Locations.Length > 0 ? Locations[0] : null;
    }

    public Location FirstLocationOrNone()
    {
        return TryGetFirstLocation() ?? Location.None;
    }

    public virtual bool IsImplicitlyDeclared => false;

    public Symbol? ContainingSymbol { get; }

    public virtual AssemblySymbol? ContainingAssembly => ContainingSymbol?.ContainingAssembly;

    public virtual NamespaceSymbol? ContainingNamespace
    {
        get
        {
            for (
                var container = ContainingSymbol;
                container is not null;
                container = container.ContainingSymbol
            )
            {
                if (container is NamespaceSymbol namespaceSymbol)
                {
                    return namespaceSymbol;
                }
            }

            return null;
        }
    }

    public virtual TypeSymbol? ContainingType
    {
        get
        {
            var containingType = ContainingSymbol as TypeSymbol;
            if (ReferenceEquals(ContainingSymbol, containingType))
            {
                return containingType;
            }

            return ContainingSymbol?.ContainingType;
        }
    }

    public abstract ImmutableArray<SyntaxReference> DeclaringSyntaxReferences { get; }

    public virtual bool IsDefinedInSourceTree(SyntaxTree tree, TextSpan? definedWithin)
    {
        if (IsImplicitlyDeclared && DeclaringSyntaxReferences.IsEmpty)
        {
            Debug.Assert(ContainingSymbol is not null);
            return ContainingSymbol.IsDefinedInSourceTree(tree, definedWithin);
        }

        // ReSharper disable once ForeachCanBeConvertedToQueryUsingAnotherGetEnumerator
        foreach (var reference in DeclaringSyntaxReferences)
        {
            if (IsDefinedInSourceTree(reference, tree, definedWithin))
                return true;
        }

        return false;
    }

    internal static bool IsDefinedInSourceTree(
        SyntaxReference reference,
        SyntaxTree tree,
        TextSpan? definedWithin
    )
    {
        return ReferenceEquals(reference.SyntaxTree, tree)
            && (definedWithin is null || reference.SourceSpan.IntersectsWith(definedWithin.Value));
    }

    public abstract void WriteDisplayString(TextWriter writer);

    public string ToDisplayString()
    {
        var writer = new StringWriter();
        WriteDisplayString(writer);
        return writer.ToString();
    }

    internal virtual Compilation? DeclaringCompilation => ContainingAssembly?.DeclaringCompilation;

    private protected void AddDeclarationDiagnostics(DiagnosticBag diagnostics)
    {
        var compilation = DeclaringCompilation;
        Debug.Assert(compilation is not null);
        compilation.DeclarationDiagnostics.AddRange(diagnostics);
    }

    internal virtual bool NeedsCompletion => false;

    internal virtual void ForceComplete(
        SourceLocation? location,
        Predicate<Symbol>? filter,
        CancellationToken cancellationToken
    )
    {
        Debug.Assert(!NeedsCompletion, $"Need to override {nameof(ForceComplete)}");
    }

    internal virtual bool IsComplete(CompletionPart part)
    {
        Debug.Assert(!NeedsCompletion, $"Need to override {nameof(IsComplete)}");
        return true;
    }

    private protected static void ForceCompleteMemberConditionally(
        SourceLocation? location,
        Predicate<Symbol>? predicate,
        Symbol member,
        CancellationToken cancellationToken
    )
    {
        if (
            location is null
            || member.IsDefinedInSourceTree(location.SyntaxTree, location.SourceSpan)
                && predicate?.Invoke(member) != false
        )
        {
            member.ForceComplete(location, predicate, cancellationToken);
        }
    }
}
